---
title: "Analysis of Genetic Data 2: Mapping Genome-Wide Associations"
author: Peter Carbonetto
output:
  beamer_presentation:
    template: beamer.tex
    keep_tex: false
---

```{r knitr-options, echo=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,fig.align = "center",
                      results = "hide", fig.show = "hide", message = FALSE,
		      warning = FALSE)
```

Aims of workshop
================

*Describe aims here.*

Our research task
=================

*Describe research task here.*

It's your choice
================

Your may choose to...

+ Work through the examples on the RCC cluster.

+ Work through the examples on your laptop.

+ Pair up with your neighbour.

+ Follow what I do on the projector.

*However*,

1. A few of the examples will only run on the RCC cluster.

2. The examples may not produce the exact same result on your laptop.

Software tools we will use today
================================

1. PLINK

2. R

3. R packages `ggplot2` and `cowplot`.

4. Basic shell commands such as ...

Outline of workshop
===================

*Add outline here.*

Initial setup (part 1)
======================

+ WiFi

+ Power outlets

+ YubiKeys

+ Pace, questions (e.g., keyboard shortcuts).

Initial setup (part 2)
======================

Download the workshop packet to your laptop.

+ URL: https://github.com/rcc-uchicago/genetic-data-analysis-2

+ Open **slides.pdf** from the **docs** folder. This is useful for
viewing and copying the code from the slides.

+ We may also look at some of the files in the **code** folder. The
code can be viewed with your favourite text editor, or browsed
on GitHub.

Initial setup (part 3)
======================

If you are using the RCC cluster, set up your cluster computing
environment:

1. Connect to midway2.

    + See https://rcc.uchicago.edu/docs/connecting.
    
2. *Add screen and sinteractive commands here.*

Initial setup (part 4)
======================

If you are using the RCC cluster, also download the workshop packet
to your home directory on the cluster.

* URL: https://github.com/rcc-uchicago/genetic-data-analysis-2

You can run these commands to download the workshop packet:

```{bash download-repo, eval=FALSE}
cd ~
git clone https://github.com/rcc-uchicago/
  genetic-data-analysis-2.git
```

What's in the workshop packet
=============================

```
genetic-data-analysis-2
  /bin     # All executables are stored here.
  /code    # Source code used in analyses.
  /data    # Raw and processed data.
  /docs    # Additional workshop materials.
  /output  # All results are stored here.
```

Download data (part 1)
======================

The genotype files are too large to fit in the git repository. So we
need to download them separately.

+ URL: http://mtweb.cs.ucl.ac.uk/dosages

+ Download the files for chromosomes 1 through 19 into the `data`
  folder.

+ You may find it easier to run the bash script provided in the `data`
  folder:

    ```
    bash download_genotypes.sh
    ```

    This script requires `wget`.

Download data (part 2)
======================

After following these steps, this should be the contents of your
`data` directory:

```
$ ls -1
README
CFW_covariates.txt
CFW_measures.txt
download_genotypes.sh
chr1.prunedgen.final.maf001.0.98.RData
chr2.prunedgen.final.maf001.0.98.RData
...
chr19.prunedgen.final.maf001.0.98.RData
```

The data we will use in our analyses
====================================

In the `data` folder:

+ **CFW_measures.txt:** 200 phenotypes collected from 2,117 CFW mice.

+ **CFW_covariates.txt:** covariate data for 2,117 CFW mice.

+ **list_of_1934_mice_used_for_analysis.txt:** ids of the 1,934 mice
  used in the final analyses of Nicod *et al*, 2016.

See the `README` in the `data` folder for information.

Explore the data from the shell
===============================

First take a quick look the data using command-line tools:

```{bash, eval=FALSE}
less -S CFW_measures.txt
wc -l CFW_measures.txt
```

```{bash, eval=FALSE}
less -S CFW_covariates.txt
wc -l CFW_covariates.txt
```

Explore the data in R
=====================

Load the phenotype and covariate data, and merge these data into a
single table ("data frame"):

```{r}
dat1  <- read.flint.pheno("../data/CFW_measures.txt")
dat2  <- read.flint.pheno("../data/CFW_covariates.txt")
pheno <- cbind(dat1,dat2[-1])
```

Take a quick look at the data:

```{r}
nrow(pheno)
ncol(pheno)
pheno[1:5,1:5]
```

Select the same 1,934 samples that were used in Nicod et al (2016);

```{r}
ids   <- read.table("../data/list_of_1934_mice_used_for_analysis.txt",
                    stringsAsFactors = FALSE)[[1]]
rows  <- match(ids,pheno$Animal_ID)
pheno <- pheno[rows,]
nrow(pheno)
```

Take a closer look at the tibia length data (units = mm):

```{r}
summary(pheno$Tibia.Length)
library(ggplot2)
library(cowplot) # Optional.
p1 <- plot.dist(pheno$Tibia.Length)
print(p1)
p1 <- p1 + labs(x = "tibia length (mm)",y = "number of mice")
print(p1)
```

Examine the relationship between tibia length and body weight:

```{r}
```

*Optional exercise:* Follow these same processing and analysis steps
to map genetic loci for LDL concentrations in blood (units = mmol/L).

