---
title: "Analysis of Genetic Data 2: Mapping Genome-Wide Associations"
author: Peter Carbonetto
output:
  beamer_presentation:
    template: beamer.tex
    keep_tex: false
---

```{r knitr-options, echo=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,fig.align = "center",
                      results = "hide", fig.show = "hide", message = FALSE,
		      warning = FALSE)
```

Aims of workshop
================

*Describe aims here.*

Our research task
=================

*Describe research task here.*

It's your choice
================

Your may choose to...

+ Work through the examples on the RCC cluster.

+ Work through the examples on your laptop.

+ Pair up with your neighbour.

+ Follow what I do on the projector.

*However*,

1. A few of the examples will only run on the RCC cluster.

2. The examples may not produce the exact same result on your laptop.

Software tools we will use today
================================

1. GEMMA

2. R

3. R packages `stringr`, `ggplot2` and `cowplot`.

4. Basic shell commands such as `wc` and `cat`.

Outline of workshop
===================

*Add outline here.*

Initial setup (part 1)
======================

+ WiFi

+ Power outlets

+ YubiKeys

+ Pace, questions (e.g., keyboard shortcuts).

Initial setup (part 2)
======================

Download the workshop packet to your laptop.

+ URL: https://github.com/rcc-uchicago/genetic-data-analysis-2

+ Open **slides.pdf** from the **docs** folder. This is useful for
viewing and copying the code from the slides.

+ We may also look at some of the files in the **code** folder. The
code can be viewed with your favourite text editor, or browsed
on GitHub.

Initial setup (part 3)
======================

If you are using the RCC cluster, set up your cluster computing
environment:

1. Connect to midway2.

    + See https://rcc.uchicago.edu/docs/connecting.

    + If you are using ThinLinc, use the "clipboard" to copy & paste.
    
2. *Add screen and sinteractive commands here.*

Initial setup (part 4)
======================

If you are using the RCC cluster, also download the workshop packet
to your home directory on the cluster.

* URL: https://github.com/rcc-uchicago/genetic-data-analysis-2

You can run these commands to download the workshop packet:

```{bash download-repo, eval=FALSE}
cd ~
git clone https://github.com/rcc-uchicago/
  genetic-data-analysis-2.git
```

What's in the workshop packet
=============================

```
genetic-data-analysis-2
  /bin     # All executables are stored here.
  /code    # Source code used in analyses.
  /data    # Raw and processed data.
  /docs    # Additional workshop materials.
  /output  # All results are stored here.
```

Download data (part 1)
======================

The genotype files are too large to fit in the git repository. So we
need to download them separately.

+ URL: http://mtweb.cs.ucl.ac.uk/dosages

+ Download the files for chromosomes 1 through 19 into the `data`
  folder.

+ You may find it easier to run the bash script provided in the `data`
  folder:

    ```
	cd data
    bash download_genotypes.sh
    ```

    This script requires the `wget` program.

Download data (part 2)
======================

After following these steps, this should be the contents of your
`data` directory:

```
$ cd data
$ ls -1
README
CFW_covariates.txt
CFW_measures.txt
download_genotypes.sh
chr1.prunedgen.final.maf001.0.98.RData
chr2.prunedgen.final.maf001.0.98.RData
...
chr19.prunedgen.final.maf001.0.98.RData
```

The data we will use in our analyses
====================================

Key files in the `data` folder:

+ **CFW_measures.txt:** 200 phenotypes collected from 2,117 CFW mice.

+ **CFW_covariates.txt:** covariate data for 2,117 CFW mice.

+ **listof1934miceusedforanalysis.txt:** ids of the 1,934 mice
  used in the final analyses of Nicod *et al*, 2016.

+ **chr\*.prunedgen.final.maf001.0.98.RData:** genotypes for SNPs on
  chromosomes 1--19 used in the final analysis of Nicod *et al*, 2016.

See the `README` in the `data` folder for more details.

Explore the phenotype data from the terminal
============================================

First, let's take a quick look the phenotype data using command-line
tools:

```{bash, eval=FALSE}
less -S CFW_measures.txt
wc -l CFW_measures.txt
```

Likewise, the covariate data:

```{bash, eval=FALSE}
less -S CFW_covariates.txt
wc -l CFW_covariates.txt
```

Explore the phenotype data in R (part 1)
========================================

Move to the **code** folder in the git repository. And on the RCC
cluster, load R:

```{bash, eval=FALSE}
cd code
module load R
```

Start up an interactive R environment:

```{bash, eval=FALSE}
R --no-save
```

Check which version of R you are running (hopefully version 3.4.0 or
greater):

```{r check-r-version}
version$version.string
```

Also, before continuing, check your working directory:

```{r check-wd}
getwd()  # Should be .../code
```

Explore the phenotype data in R (part 2)
========================================

Load the phenotype and covariate data, and merge these data into a
single table ("data frame"), dropping the redundant second "id"
column:

```{r load-pheno-data}
source("functions.R")
dat1 <- read.pheno("../data/CFW_measures.txt")
dat2 <- read.pheno("../data/CFW_covariates.txt")
pheno <- cbind(dat1,dat2[-1])
```

Take a quick glance at the data:

```{r glance-at-pheno-data}
nrow(pheno)
ncol(pheno)
sort(names(pheno))
pheno[1:5,1:5]
```

Explore the phenotype data in R (part 3)
========================================

Select the same 1,934 samples that were used in Nicod *et al*
(2016). First, we load the list of ids:

```{r load-ids}
x <- "../data/listof1934miceusedforanalysis.txt"
ids <- read.table(x,stringsAsFactors = FALSE)
ids <- ids[[1]]
```

Then we select the rows of the table that match up with the ids:

```{r select-ids}
rows  <- match(ids,pheno$Animal_ID)
pheno <- pheno[rows,]
nrow(pheno)
```

Explore the phenotype data in R (part 4)
========================================

Take a closer look at the tibia length data (units = mm):

```{r summarize-tibia-length}
summary(pheno$Tibia.Length)
```

Plot the distribution of tibia length (`plot.dist` was defined in
`functions.R`):

```{r plot-tibia-length-dist}
library(ggplot2)
library(cowplot) # Optional.
p1 <- plot.dist(pheno$Tibia.Length)
print(p1)
```

Fix the axis labels, following Best Practices:

```{r plot-tibia-length-dist-2}
p1 <- p1 + labs(x = "tibia length (mm)",
                y = "number of mice")
print(p1)
```

Explore the covariate data in R (part 1)
========================================

The authors recommend using body weight as a covariate in the
association analysis:

```{r summarize-body-weight}
summary(pheno$Weight.Diss)
```

Examine the relationship between tibia length and body weight:

```{r plot-tibia-vs-weight}
p2 <- draw.scatterplot(pheno$Weight.Diss,
                       pheno$Tibia.Length) +
  labs(x = "body weight (g)",
       y = "tibia length (mm)")
print(p2)
```

Explore the covariate data in R (part 2)
========================================

The authors also recommend using sex as a covariate in the association
analysis:

```{r summarize-sex}
pheno <- transform(pheno,Sex = factor(Sex))
summary(pheno$Sex)
```

A boxplot is the standard approach to visualize the relationship
between a continuous variable (tibia length) and a categorical
variable (sex):

```{r plot-tibia-vs-sex}
p3 <- draw.boxplot(pheno$Sex,
                   pheno$Tibia.Length) +
  labs(x = "sex",y = "tibia length (mm)")
print(p3)
```

Verify the phenotype residuals
==============================

Compute the phenotypes after removing the linear effects of sex and
body weight:

```{r compute-residuals}
fit <- lm(Tibia.Length ~ Sex + Weight.Diss,pheno)
r   <- resid(fit)
```

Plot the distribution of the residuals:

```{r plot-residuals}
plot.dist(r) + labs(x = "residuals")
```

Is the normal distribution a good fit for the residuals? We can verify
this more rigorously:

```{r check-residuals}
check.normal.quantiles(r)
draw.qqplot(r)
```

Save phenotype data for GEMMA
=============================

Now that we have explored and verified the phenotype data, we are
ready to save the data in a format usable by GEMMA. First write the
phenotype data to `pheno.txt`:

```{r write-phenotype-data}
write.table(pheno$Tibia.Length,
            "../data/pheno.txt",
            row.names = FALSE,
			col.names = FALSE)
```

Save covariate data for GEMMA
=============================

Write the covariate data to file `covariates.txt` in the format used
by GEMMA. This requires that we convert sex to a numeric value:

```{r write-covariate-data}
sex <- as.numeric(pheno$Sex) - 1
unique(sex)
write.table(data.frame(1,sex,pheno$Weight.Diss),
            "../data/covariates.txt",sep = " ",
			row.names = FALSE,col.names = FALSE)
```

Also, a column of ones is needed for an intercept in the GEMMA
regression analysis.

Prepare the genotype data for GEMMA (part 1)
============================================

Processing the genotype data is more complicated, so I have provided a
script to create, for each chromosome, a SNP annotation file and a
genotype file in the BIMBAM format. Run this script in R:

```{r write-genotype-data, eval=FALSE}
source("format.genotypes.for.gemma.R")
```

+ This will take a few minutes to run. Once it is done, run `q()` to
exit R.

Let's look at the GEMMA files:

```{bash inspect-gemma-files, eval=FALSE}
less chr01.map.txt
wc -l chr01.map.txt
less -S chr01.geno.txt
wc -l chr01.geno.txt
```

*What do the numeric values in* `chr01.geno.txt` *represent?*

Prepare the genotype data for GEMMA (part 2)
============================================

The genotype data need to be combined into one file. This can be
easily done with the `cat` command, which combines by lines (rows):

```{bash merge-genotype-data, eval=FALSE}
cat chr*.geno.txt > geno.txt
head geno.txt | less -S
tail geno.txt | less -S
wc -l geno.txt
```

We do the same for the SNP annotations:

```{bash merge-snp-data, eval=FALSE}
cat chr*.map.txt > map.txt
wc -l map.txt
```

*Why didn't we combine the genotypes inside the R script?*

Prepare the genotype data for GEMMA (part 3)
============================================

A **critical step** in this script was to make sure that the samples
in the genotype and phenotype files are listed in the same order. *A
common mistake is to save the phenotype and genotype samples in
different orders, which will lead to an incorrect association
analysis!*

+ Why is it easy to make this mistake in a GEMMA analysis?

Install GEMMA
=============

Download the GEMMA 0.96 to the `bin` folder.

+ URL: https://github.com/genetics-statistics/GEMMA/releases

On the RCC cluster, you can run these commands to install the GEMMA
0.96 Linux binary, and check that it works:

```{bash get-gemma, eval=FALSE}
cd bin
wget http://bit.ly/2I67fBV -O gemma.linux.gz
gunzip gemma.gz
mv gemma.linux gemma
chmod 700 gemma
./gemma
```

Basic association analysis in GEMMA (part 1)
============================================

Now that we have installed GEMMA, and we have prepared the phenotype
and genotype data in the formats accepted by GEMMA, we are now ready
to run our first association analysis. 

```{bash gemma-lm, eval=FALSE}
cd data
../bin/gemma -p pheno.txt -c covariates.txt \
  -a map.txt -g geno.txt -notsnp -lm 2 \
  -outdir ../output -o tibia
```

The `-notsnp` option is included here to skip any additional
processing of the genotype data.

Basic association analysis in GEMMA (part 2)
============================================

This GEMMA command creates two files:

+ **tibia.log.txt:** A brief summary of the association analysis.

+ **tibia.assoc.txt:** The full table of association results
  (*p*-values, effect size estimates, *etc*).

Use `less -S` or your favourite text editor to inspect these files.

Assess "genomic inflation" (part 1)
===================================

A simple q-q plot is commonly used to assess inflation of small
$p$-values, which could indicate a unacceptably high rate of false
positive associations. This is a very rudimentary test, and has
several problems, but is nonetheless useful as a simple
diagnostic. Here we create a q-q plot in R.

```{bash, eval=FALSE}
cd code
R --no-save
```

Next, use `read.table` to load the results of the GEMMA association
analysis:

```{r load-gemma-pvalues}
gwscan <- read.table("../output/tibia.assoc.txt",
                     as.is = "rs",header = TRUE)
head(gwscan)
nrow(gwscan)
```

Assess "genomic inflation" (part 2)
===================================

Plot the observed *p*-values against the expected *p*-values under the
null distribution:

```{r plot-inflation}
library(ggplot2)
library(cowplot) # Optional.
source("functions.R")
plot.inflation(gwscan$p_lrt)
```

Next, we will attempt an LMM-based association analysis, which can
reduce inflation due to population structure and "hidden"relatedness.
Save the current plot because we will compare against it, then quit R.

LMM association analysis in GEMMA (part 1)
==========================================

First, generate a "realized relatedness" matrix from the (centered)
genotypes. This may take 5-10 minutes.

```{bash gemma-kinship, eval=FALSE}
cd data
../bin/gemma -p pheno.txt -g geno.txt \
  -a map.txt -gk 1 -outdir ../output \
  -o tibia
less -S ../output/tibia.cXX.txt
wc -l ../output/tibia.cXX.txt
```

Next, fit a linear-mixed model separately for each SNP:

```{bash gemma-lmm, eval=FALSE}
../bin/gemma -p pheno.txt -c covariates.txt \
  -a map.txt -g geno.txt -notsnp \
  -k ../output/tibia.cXX.txt \
  -lmm 2 -outdir ../output -o tibia-lmm
```

This may take over 30 minutes, so I have provided a compressed file in
the `output` directory containing the LMM-based association results.

Assess "genomic inflation" in LMM analysis (part 1)
===================================================

Start up R again:

```{bash, eval=FALSE}
cd code
R --no-save
```

Follow the same steps as before to load the results of the GEMMA
association analysis, and create the genomic inflation plot:

```{r plot-gemma-lmm-pvalues}
library(ggplot2)
library(cowplot) # Optional.
source("functions.R")
gwscan <- read.table("../output/tibia-lmm.assoc.txt",
                     as.is = "rs",header = TRUE)
plot.inflation(gwscan$p_lrt)
```

Compare against the previous inflation plot---*did the LMM reduce
inflation of p-values?*

Additional slide
================

*Optional exercise (shown in a separate slide):* Follow these same
processing and analysis steps to map genetic loci for LDL
concentrations in blood (units = mmol/L).

