---
title: "Analysis of Genetic Data 2: Mapping Genome-Wide Associations"
author: Peter Carbonetto
output:
  beamer_presentation:
    template: beamer.tex
    keep_tex: false
---

```{r knitr-options, echo=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,fig.align = "center",
                      results = "hide", fig.show = "hide", message = FALSE,
		      warning = FALSE)
```

Aims of workshop
================

*Describe aims here.*

Our research task
=================

*Describe research task here.*

It's your choice
================

Your may choose to...

+ Work through the examples on the RCC cluster.

+ Work through the examples on your laptop.

+ Pair up with your neighbour.

+ Follow what I do on the projector.

*However*,

1. A few of the examples will only run on the RCC cluster.

2. The examples may not produce the exact same result on your laptop.

Software tools we will use today
================================

1. PLINK

2. R

3. R packages `ggplot2` and `cowplot`.

4. Basic shell commands such as ...

Outline of workshop
===================

*Add outline here.*

Initial setup (part 1)
======================

+ WiFi

+ Power outlets

+ YubiKeys

+ Pace, questions (e.g., keyboard shortcuts).

Initial setup (part 2)
======================

Download the workshop packet to your laptop.

+ URL: https://github.com/rcc-uchicago/genetic-data-analysis-2

+ Open **slides.pdf** from the **docs** folder. This is useful for
viewing and copying the code from the slides.

+ We may also look at some of the files in the **code** folder. The
code can be viewed with your favourite text editor, or browsed
on GitHub.

Initial setup (part 3)
======================

If you are using the RCC cluster, set up your cluster computing
environment:

1. Connect to midway2.

    + See https://rcc.uchicago.edu/docs/connecting.
    
2. *Add screen and sinteractive commands here.*

Initial setup (part 4)
======================

If you are using the RCC cluster, also download the workshop packet
to your home directory on the cluster.

* URL: https://github.com/rcc-uchicago/genetic-data-analysis-2

You can run these commands to download the workshop packet:

```{bash download-repo, eval=FALSE}
cd ~
git clone https://github.com/rcc-uchicago/
  genetic-data-analysis-2.git
```

What's in the workshop packet
=============================

```
genetic-data-analysis-2
  /bin     # All executables are stored here.
  /code    # Source code used in analyses.
  /data    # Raw and processed data.
  /docs    # Additional workshop materials.
  /output  # All results are stored here.
```

Download data (part 1)
======================

The genotype files are too large to fit in the git repository. So we
need to download them separately.

+ URL: http://mtweb.cs.ucl.ac.uk/dosages

+ Download the files for chromosomes 1 through 19 into the `data`
  folder.

+ You may find it easier to run the bash script provided in the `data`
  folder:

    ```
    bash download_genotypes.sh
    ```

    This script requires `wget`.

Download data (part 2)
======================

After following these steps, this should be the contents of your
`data` directory:

```
$ ls -1
README
CFW_covariates.txt
CFW_measures.txt
download_genotypes.sh
chr1.prunedgen.final.maf001.0.98.RData
chr2.prunedgen.final.maf001.0.98.RData
...
chr19.prunedgen.final.maf001.0.98.RData
```

The data we will use in our analyses
====================================

In the `data` folder:

+ **CFW_measures.txt:** 200 phenotypes collected from 2,117 CFW mice.

+ **CFW_covariates.txt:** covariate data for 2,117 CFW mice.

+ **listof1934miceusedforanalysis.txt:** ids of the 1,934 mice
  used in the final analyses of Nicod *et al*, 2016.

+ **chr\*.prunedgen.final.maf001.0.98.RData:** imputed genotypes for
  SNPs on chromosomes 1--19 used in the final analysis of Nicod *et
  al*, 2016.

See the `README` in the `data` folder for more details.

Explore the phenotype data from the terminal
============================================

First, let's take a quick look the phenotype data using command-line
tools:

```{bash, eval=FALSE}
less -S CFW_measures.txt
wc -l CFW_measures.txt
```

Likewise, the covariate data:

```{bash, eval=FALSE}
less -S CFW_covariates.txt
wc -l CFW_covariates.txt
```

Explore the phenotype data in R (part 1)
========================================

Move to the **code** folder in the git repository. And on the RCC
cluster, load R:

```{bash, eval=FALSE}
cd code
module load R
```

Start up an interactive R environment:

```{bash, eval=FALSE}
R --no-save
```

Check which version of R you are running (hopefully version 3.4.0 or
greater):

```{r check-r-version}
version$version.string
```

Also, before continuing, check your working directory:

```{r check-wd}
getwd()  # Should be .../code
```

Explore the phenotype data in R (part 2)
========================================

Load the phenotype and covariate data, and merge these data into a
single table ("data frame"), dropping the redundant second "id"
column:

```{r load-pheno-data}
source("functions.R")
dat1 <- read.pheno("../data/CFW_measures.txt")
dat2 <- read.pheno("../data/CFW_covariates.txt")
pheno <- cbind(dat1,dat2[-1])
```

Take a quick glance at the data:

```{r glance-at-pheno-data}
nrow(pheno)
ncol(pheno)
sort(names(pheno))
pheno[1:5,1:5]
```

Explore the phenotype data in R (part 3)
========================================

Select the same 1,934 samples that were used in Nicod *et al*
(2016). First, we load the list of ids:

```{r load-ids}
x <- "../data/listof1934miceusedforanalysis.txt"
ids <- read.table(x,stringsAsFactors = FALSE)
ids <- ids[[1]]
```

Then we select the rows of the table that match up with the ids:

```{r select-ids}
rows  <- match(ids,pheno$Animal_ID)
pheno <- pheno[rows,]
nrow(pheno)
```

Explore the phenotype data in R (part 4)
========================================

Take a closer look at the tibia length data (units = mm):

```{r summarize-tibia-length}
summary(pheno$Tibia.Length)
```

Plot the distribution of tibia length (`plot.dist` was defined in
`functions.R`):

```{r plot-tibia-length-dist}
library(ggplot2)
library(cowplot) # Optional.
p1 <- plot.dist(pheno$Tibia.Length)
print(p1)
```

Fix the axis labels in the plot:

```{r plot-tibia-length-dist-2}
p1 <- p1 + labs(x = "tibia length (mm)",
                y = "number of mice")
print(p1)
```

Explore the covariate data in R (part 1)
========================================

The authors recommend using body weight as a covariate in the
association analysis:

```{r summarize-body-weight}
summary(pheno$Weight.Diss)
```

Examine the relationship between tibia length and body weight:

```{r plot-tibia-vs-weight}
p2 <- draw.scatterplot(pheno$Weight.Diss,
                       pheno$Tibia.Length) +
  labs(x = "body weight (g)",
       y = "tibia length (mm)")
print(p2)
```

Explore the covariate data in R (part 2)
========================================

The authors also recommend using sex as a covariate in the association
analysis:

```{r summarize-sex}
pheno <- transform(pheno,Sex = factor(Sex))
summary(pheno$Sex)
```

Examine the relationship between tibia length and sex:

```{r plot-tibia-vs-sex}
p3 <- draw.boxplot(pheno$Sex,
                   pheno$Tibia.Length) +
  labs(x = "sex",y = "tibia length (mm)")
print(p3)
```

Verify the phenotype residuals
==============================

Compute the phenotypes after removing the linear effects of sex and
body weight:

```{r compute-residuals}
fit <- lm(Tibia.Length ~ Sex + Weight.Diss,pheno)
r   <- resid(fit)
```

Plot the distribution of the residuals:

```{r plot-residuals}
p4 <- plot.dist(r) + labs(x = "residuals")
print(p4)
```

Is the normal a good fit for the residuals? We can verify this more
rigorously:

```{r check-residuals}
check.normal.quantiles(r)
p5 <- draw.qqplot(r)
print(p5)
```

Save phenotype data for GEMMA
=============================

Now that we have explored and verified the phenotype data, we are
ready to save the data in a format usable by GEMMA. First write the
phenotype data to `pheno.txt`:

```{r write-phenotype-data}
write.table(pheno$Tibia.Length,"../data/pheno.txt",
            row.names = FALSE,col.names = FALSE)
```

Save covariate data for GEMMA
=============================

Write the covariate data to file `cov.txt` in the format used by
GEMMA. This requires that we convert sex to a numeric value:

```{r write-covariate-data}
sex <- as.numeric(pheno$Sex) - 1
write.table(data.frame(1,sex,pheno$Weight.Diss),
            "../data/covariates.txt",sep = " ",
			row.names = FALSE,col.names = FALSE)
```

Also, a column of ones is needed for the intercept in GEMMA.

Prepare the genotype data for GEMMA
===================================


Additional slide
================

*Optional exercise (shown in a separate slide):* Follow these same
processing and analysis steps to map genetic loci for LDL
concentrations in blood (units = mmol/L).

